src/NuGet.Core/NuGet.Packaging/PackageExtraction/StreamExtensions.cs

```diff
@@ -27,12 +34,24 @@ public static string CopyToFile(this Stream inputStream, string fileFullPath)
                 return fileFullPath;
             }
 
+            // For files of a certain size, we can do some Cleverness and mmap them instead
+            // of writing directly to disk. This can increase performance by a LOT, at the
+            // cost of files taking longer to get to disk -- but that's fine!
+            if (size > 0 && size <= MAX_MMAP_SIZE)
+            {
+                using (MemoryMappedFile mmf = MemoryMappedFile.CreateFromFile(fileFullPath, FileMode.Create, null, (long)size))
+                {
+                    MemoryMappedViewStream mmstream = mmf.CreateViewStream();
+                    inputStream.CopyTo(mmstream);
```

Comment:
I think the size above will be rounded up to fit evenly with memory page size, so the file can actually be larger than the size you just wrote. You'll want to to follow up to to fix up the size to the data you *just wrote*. To do this, I think you need to use FileStream overload and call SetLength on it after you've written the data. That might end up losing some of the benefit - I'm not sure.

Reply:
So I looked into this, and it seems like CreateFromFile/CreateViewStream do the right thing here.

I tested with:
![image](https://user-images.githubusercontent.com/17535/87813193-4921a980-c816-11ea-824f-c71bdc1710f9.png)

and the result:
![image](https://user-images.githubusercontent.com/17535/87813229-5c347980-c816-11ea-97e2-5b499089bcd9.png)

Reply:
Yeah I did a quick experiment and familized myself with the source code again, and I think you are right - it seems to factor this when creating the stream.